File: algorithms/java/src/test/java/leetcode/LC038CountAndSay.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/count-and-say/
 * #medium
 */
public class LC038CountAndSay {
    public final class Solution {
        public String countAndSay(int n) {
            if (n < 1) return null;
            String result = "1";
            while (n > 1) {
                StringBuilder current = new StringBuilder();
                char check = '0';
                int count = 0;
                for (char character : result.toCharArray()) {
                    if (check != character) {
                        if (count > 0) {
                            current.append(count);
                            current.append(check);
                        }
                        count = 1;
                        check = character;
                    } else {
                        count++;
                    }
                }
                if (count > 0) {
                    current.append(count);
                    current.append(check);
                }
                n--;
                result = current.toString();
            }
            return result;
        }
    }

    @Test
    public void test_1() throws Exception {
        assertEquals("1", new Solution().countAndSay(1));
    }

    @Test
    public void test_2() throws Exception {
        assertEquals("11", new Solution().countAndSay(2));
    }

    @Test
    public void test_3() throws Exception {
        assertEquals("21", new Solution().countAndSay(3));
    }

    @Test
    public void test_4() throws Exception {
        assertEquals("1211", new Solution().countAndSay(4));
    }

    @Test
    public void test_5() throws Exception {
        assertEquals("111221", new Solution().countAndSay(5));
    }
}

File: algorithms/code/leetcode/lc038_count_and_say/lc038_count_and_say.py
# https://leetcode.com/problems/count-and-say/

import unittest


class Solution:
    def countAndSay(self, n: int) -> str:
        if n < 1:
            return ''
        result = '1'
        while n > 1:
            current = ''
            check = '0'
            count = 0
            for character in result:
                if check != character:
                    if count > 0:
                        current += str(count)
                        current += check
                    count = 1
                    check = character
                else:
                    count += 1
            current += str(count)
            current += check
            n -= 1
            result = current
        return result


class TestCode(unittest.TestCase):
    def test_1(self) -> None:
        assert Solution().countAndSay(1) == '1'

    def test_2(self) -> None:
        assert Solution().countAndSay(2) == '11'

    def test_3(self) -> None:
        assert Solution().countAndSay(3) == '21'

    def test_4(self) -> None:
        assert Solution().countAndSay(4) == '1211'

    def test_5(self) -> None:
        assert Solution().countAndSay(5) == '111221'

    def test_0(self) -> None:
        assert Solution().countAndSay(0) == ''
